extensions/amp-fresh/0.1/amp-fresh-manager.js

```diff
@@ -14,14 +14,98 @@
  * limitations under the License.
  */
 
+import {addParamToUrl} from '../../../src/url';
 import {fromClass} from '../../../src/service';
+import {isExperimentOn} from '../../../src/experiments';
+import {whenDocumentReady} from '../../../src/document-ready';
+import {xhrFor} from '../../../src/xhr';
 
 
+/** @const */
+const TAG = 'amp-fresh';
+
 /**
  * Manages registered amp-fresh components and does the xhr request
  * and update.
  */
-export class AmpFreshManager {}
+export class AmpFreshManager {
+
+  /**
+   * @param {!Window} win
+   */
+  constructor(win) {
+
+    /** @type {!Window} */
+    this.win = win;
+
+    if (!isExperimentOn(this.win, TAG)) {
+      return;
+    }
+
+    /** @private @const {!Object<string, !./amp-fresh.AmpFresh>} */
+    this.ampFreshInstances_ = Object.create(null);
+
+    /** @private {?Document} */
+    this.docPromise_ = null;
+
+    this.fetchDocument_().then(doc => {
+      this.update_(doc);
+    });
+  }
+
+  /**
+   * Registers an amp-fresh component instance to be managed by this windows
+   * amp-fresh-manager instance.
+   * @param {string} id
+   * @param {!./amp-fresh.AmpFresh} ampFreshIntance
+   */
+  register(id, ampFreshInstance) {
+    this.ampFreshInstances_[id] = ampFreshInstance;
+  }
+
+  /**
+   * @return {!Promise<!Document>}
+   */
+  fetchDocument_() {
+    if (this.docPromise_) {
+      return;
+    }
+    const url = addParamToUrl(this.win.location.href, 'amp-fresh', '1');
+    this.docPromise_ = Promise.all([
+      xhrFor(this.win).fetchDocument(url),
+      whenDocumentReady(this.win.document),
+    ]).then(args => args[0]);
+
+    this.docPromise_.catch(() => {
```

Comment:
so i wanted to propagate the error, but also make sure that we unhid all the elements even on a failure.

 what do you think?